### PR TITLE
Merge https and http3 reverse-proxy modes

### PR DESCRIPTION
**Pull request by errorxyz**
#### Description

Merge http3 reverse proxy mode into https mode so that we listen for http3 connections as well while in https mode

#### Checklist

 - [x] I have updated tests where applicable.
 - [ ] I have added an entry to the CHANGELOG.

### CODE REVIEW

File: mitmproxy/addons/next_layer.py [hunk 1]
```diff
@@ -350,10 +350,15 @@ def _setup_reverse_proxy(context: Context, data_client: bytes) -> Layer:
                     stack /= ClientTLSLayer(context)
                 stack /= HttpLayer(context, HTTPMode.transparent)
             case "https":
-                stack /= ServerTLSLayer(context)
-                if starts_like_tls_record(data_client):
-                    stack /= ClientTLSLayer(context)
-                stack /= HttpLayer(context, HTTPMode.transparent)
+                if _starts_like_quic(data_client):
```
Comment (mhils) [inline mitmproxy/addons/next_layer.py:353]: Can we inspect client_conn.transport_protocol instead? That feels way more robust.

File: mitmproxy/addons/proxyserver.py [hunk 2]
```diff
@@ -259,14 +259,21 @@ def configure(self, updated) -> None:
...
-                )
-                for m in modes
-            ]
+            listen_addrs = []
+            for m in modes:
+                protocols = []
+                if m.transport_protocol == "both":
+                    protocols.extend(["tcp", "udp"])
+                else:
+                    protocols.append(m.transport_protocol)
```
Comment (mhils) [inline mitmproxy/addons/proxyserver.py:268]:
```suggestion
                if m.transport_protocol == "both":
                    protocols = ["tcp", "udp"]
                else:
                    protocols = [m.transport_protocol]
```

File: mitmproxy/proxy/mode_specs.py [hunk 3]
```diff
@@ -223,17 +223,15 @@ class ReverseMode(ProxyMode):
 
     description = "reverse proxy"
     transport_protocol = TCP
-    scheme: Literal[
-        "http", "https", "http3", "tls", "dtls", "tcp", "udp", "dns", "quic"
-    ]
+    scheme: Literal["http", "https", "tls", "dtls", "tcp", "udp", "dns", "quic"]
```
Comment (mhils) [inline mitmproxy/proxy/mode_specs.py:226]: For reverse proxy mode, does it maybe make sense to keep `http3`? This way a user can explicitly specify HTTP/3 only. `quic://` forces raw QUIC instead.

(I wouldn't mind adding `http1://` and `http2://` eventually)

**Review by mhils — COMMENTED** [review]
So nice how small and easy this change is now. 🤩

File: mitmproxy/proxy/mode_specs.py (hunk 3 above)
Comment (errorxyz) [reply mitmproxy/proxy/mode_specs.py:226]: Sounds good.

File: mitmproxy/addons/proxyserver.py (hunk 2 above)
Comment (errorxyz) [reply mitmproxy/addons/proxyserver.py]: We get a mypy error with this if-else pattern
`error: List item 0 has incompatible type "Literal['tcp', 'udp'] | None"; expected "str"  [list-item]`
so I've reversed the conditions
We could instead set the variable type using `protocols: list[Literal['tcp', 'udp'] | None]`, lmk if that's better

**Comment by errorxyz** [comment]
shall we block this untill https://github.com/mitmproxy/mitmproxy/issues/7112 is fixed?

File: mitmproxy/addons/proxyserver.py (hunk 2 above)
Comment (mhils) [reply mitmproxy/addons/proxyserver.py]: Ah, thanks. The mypy complaint isn't completely unjustified, `m.transport_protocol` can be None by it's type annotation. It's only set to None for local mode, and that comes from a time where "both" was not a valid option yet.  Let's remove the `| None` part for transport_protocol, make it `both` for Local/OsProxy mode, and this error should go away? 
(If it doesn't, a `protocols: list[TransportProtocol]` line before the if/else should fix it.)

**Review by mhils — APPROVED** [review]

**Pull request merged**